node/hyperbahn/service_proxy.js

```diff
@@ -295,11 +295,54 @@ ServiceDispatchHandler.prototype.refreshServicePeer =
 function refreshServicePeer(serviceName, hostPort) {
     var self = this;
 
+    // Add hostPort to service peers, decision pending on whether to connect.
     var peer = self.getServicePeer(serviceName, hostPort);
-    peer.connect();
 
+    // Make a note not to collect this service
     var time = self.channel.timers.now();
     self.exitServices[serviceName] = time;
+
+    var minWorkerPeers = 5; // TODO config
+    var minRelayPeers = 5; // TODO config
+
+    // Obtain and sort the affine worker and relay lists.
+    var relays = Object.keys(self.egressNodes.exitsFor(serviceName));
+    relays.sort();
+    var serviceChannel = self.getOrCreateServiceChannel(serviceName);
+    var workers = serviceChannel.peers.keys();
+    workers.sort();
+
+    // Find our position within the affine relay set so we can project that
+    // position into the affine worker set.
+    var relayIndex = relays.indexOf(self.channel.hostPort);
+    if (relayIndex < 0) {
+        // This should not be possible.
+        self.logger.error('Relay find itself within the affine relays for an ad it received', {
+            serviceName: serviceName,
+            relayHostPort: self.channel.hostPort,
+            workerHostPort: hostPort
+        });
+        return;
+    }
+
+    // Compute the range of workers that this relay should be connected to.
+    var ratio = workers.length / relays.length;
+    var start = relayIndex * ratio;
+    var stop = start + Math.max(minRelayPeers, minWorkerPeers * ratio) + 0.5;
+
+    // Adjust connections as necessary.
+    for (var workerIndex = 0; workerIndex < workers.length; workerIndex++) {
+        var shouldConnect = workerIndex >= start && workerIndex < stop ||
+            workerIndex > stop && workerIndex <= start;
+        peer = self.getServicePeer(serviceName, workers[workerIndex]);
+        if (shouldConnect) {
+            peer.connect();
+        } else {
+            peer.close(noop);
```

Comment:
Since we’re talking about methods that don’t exist, in the long run, I believe the semantics we need here are: remove the peer from the subchannel’s peers, and after no subchannels retain the peer (perhaps in a reaper cycle), close all connections, dropping any pending requests on the floor. This may hurt some pending requests, but they can retry.

As an expedient, I am going to submit a version of this feature that never collects peers if they have ever been connected. This should reduce the number of connected peers, but will still leak toward fully connected over time, especially while receiving initial advertisements.